### PR TITLE
Added a flag to support retrieving parameters from different regions.

### PR DESCRIPTION
**Pull request by mcnoreau12**

### CODE REVIEW

**Comment by fredboutin** [comment]
Please add the new feature in the readme and ping documentation to review it after we do a first review of this PR.

File: src/main/java/com/coveo/configuration/parameterstore/ParameterStorePropertySourceEnvironmentPostProcessor.java [hunk 1]
```diff
@@ -29,15 +35,29 @@
...
+            if (isMultiRegionEnabled(environment)) {
+                List<String> signingRegions = getSigningRegions(environment);
+                ListIterator<String> regionsIterator = signingRegions.listIterator(signingRegions.size());
+
+                while (regionsIterator.hasPrevious()) {
+                    boolean isLast = regionsIterator.hasNext();
+                    String region = regionsIterator.previous();
+
+                    environment.getPropertySources()
+                               .addFirst(buildParameterStorePropertySource(PARAMETER_STORE_PROPERTY_SOURCE_NAME
```
Comment (jebeaudet) [inline src/main/java/com/coveo/configuration/parameterstore/ParameterStorePropertySourceEnvironmentPostProcessor.java]: add an internal method that gets a `isLast` parameter to avoir a weird ternary if(true) return false ;)

File: src/main/java/com/coveo/configuration/parameterstore/ParameterStorePropertySourceEnvironmentPostProcessor.java [hunk 2]
```diff
@@ -29,15 +35,29 @@
...
-                                                                                           environment.getProperty(PARAMETER_STORE_HALT_BOOT_CONFIGURATION_PROPERTY,
-                                                                                                                   Boolean.class,
-                                                                                                                   Boolean.FALSE))));
-            if (!environment.getProperty(PARAMETER_STORE_SUPPORT_MULTIPLE_APPLICATION_CONTEXTS_CONFIGURATION_PROPERTY,
-                                         Boolean.class,
-                                         Boolean.FALSE)) {
+            boolean haltBoot = environment.getProperty(PARAMETER_STORE_HALT_BOOT_CONFIGURATION_PROPERTY,
+                                                       Boolean.class,
+                                                       Boolean.FALSE);
+            if (isMultiRegionEnabled(environment)) {
```
Comment (jebeaudet) [inline src/main/java/com/coveo/configuration/parameterstore/ParameterStorePropertySourceEnvironmentPostProcessor.java]: I think just having multiples `signingRegions` would be enough here instead of an additional parameter no? I'm also thinking the multi-region and single-region could be in their own class for clarity purposes.

File: src/main/java/com/coveo/configuration/parameterstore/ParameterStorePropertySourceEnvironmentPostProcessor.java [hunk 3]
```diff
@@ -53,15 +73,53 @@ private boolean isParameterStorePropertySourceEnabled(ConfigurableEnvironment en
...
+    private AWSSimpleSystemsManagement buildSSMClient(String region)
+    {
+        return AWSSimpleSystemsManagementClientBuilder.standard().withRegion(region).build();
+    }
+
+    private List<String> getSigningRegions(ConfigurableEnvironment environment)
+    {
+        return Arrays.asList(environment.getProperty(PARAMETER_STORE_CLIENT_ENDPOINT_SIGNING_REGIONS_CONFIGURATION_PROPERTY,
+                                                     String[].class,
+                                                     new String[] { getSigningRegion(environment) }));
```
Comment (jebeaudet) [inline src/main/java/com/coveo/configuration/parameterstore/ParameterStorePropertySourceEnvironmentPostProcessor.java]: default on the single region? it's a weird behavior no?

File: src/main/java/com/coveo/configuration/parameterstore/ParameterStorePropertySourceEnvironmentPostProcessor.java (hunk 2 above)
Comment (mcnoreau12) [reply src/main/java/com/coveo/configuration/parameterstore/ParameterStorePropertySourceEnvironmentPostProcessor.java]: Yeah. Thought about it and it probably would be much simpler. I added the property `multiRegion.enabled` to be more flexible. Multi region could be enabled without specifying the regions and it could fallback on `signingRegion` or the default one. But it's just too flexible. You don't need to enabled multi region if you don't have multiple regions. I'll change it to just `signingRegions`. I'll also put them in separate classes.

File: src/main/java/com/coveo/configuration/parameterstore/ParameterStorePropertySourceEnvironmentPostProcessor.java [hunk 4]
```diff
@@ -29,15 +35,29 @@
...
+            boolean haltBoot = environment.getProperty(PARAMETER_STORE_HALT_BOOT_CONFIGURATION_PROPERTY,
+                                                       Boolean.class,
+                                                       Boolean.FALSE);
+            if (isMultiRegionEnabled(environment)) {
+                List<String> signingRegions = getSigningRegions(environment);
+                ListIterator<String> regionsIterator = signingRegions.listIterator(signingRegions.size());
+
+                while (regionsIterator.hasPrevious()) {
+                    boolean isLast = regionsIterator.hasNext();
+                    String region = regionsIterator.previous();
```
[the remark this answers is not in the document]
Comment (mcnoreau12) [reply src/main/java/com/coveo/configuration/parameterstore/ParameterStorePropertySourceEnvironmentPostProcessor.java]: I'm not sure it works. I want to know if the current region is the last one in the list. I'm iterating from the last one to the first one so I'm not sure I should check if there's a previous to know if it's the last no? It's too confusing I think. I'll change it to make it easier to understand.

File: readme.md [hunk 5]
```diff
@@ -81,6 +81,12 @@ Spring Cloud has a second application context named bootstrap that gets initiali
 
 If you still want the post processor to run twice or if you are using [spring-boot-devtools](https://docs.spring.io/spring-boot/docs/current/reference/html/using-spring-boot.html#using-boot-devtools-restart), you can set the optional property `awsParameterStorePropertySource.supportMultipleApplicationContexts` to `true`. The default property value is `false`to prevent multiple initializations. If you are also using Spring Cloud, this property will only work if set in the bootstrap properties.
 
+## Multi-region
+- Set `awsParameterStoreSource.ssmClient.signingRegions` with the regions from which you need to retrieve parameters using a **comma-separated** list such as `us-east-1,us-east-2`. It adds a PropertySource for each region specified. It will start looking from the first region and so on until it finds the property so put the regions in order of precedence.
```
Comment (jebeaudet) [inline readme.md]: I think we should divert from the `signingRegions` key, it's misleading IMO. This is refering to the other feature of specifying the endpoint + signingRegion and this is not the case here.

I suggest `awsParameterStoreSource.multiRegion.ssmClient.regions`. With this we'll have a new subnamespace `multiRegion` for future improvements. Thoughts?

Comment (mcnoreau12) [reply readme.md]: Yes sounds better! Thanks

File: readme.md [hunk 6]
```diff
@@ -81,6 +81,12 @@ Spring Cloud has a second application context named bootstrap that gets initiali
...
 
+## Multi-region
+- Set `awsParameterStoreSource.ssmClient.signingRegions` with the regions from which you need to retrieve parameters using a **comma-separated** list such as `us-east-1,us-east-2`. It adds a PropertySource for each region specified. It will start looking from the first region and so on until it finds the property so put the regions in order of precedence.  
+**Reminder**: using other list injecting methods like a yaml list won't work because this property gets loaded too early in the boot process.
+- If you want to halt the boot when a property isn't found in any of the specified regions, just set `awsParameterStorePropertySource.haltBoot` to `true` in your properties.
+- Make sure that your service has the necessary permissions to access parameters in the specified regions
```
[the remark this answers is not in the document]
Comment (mcnoreau12) [reply readme.md]: done

File: src/main/java/com/coveo/configuration/parameterstore/strategy/MultiRegionParameterStorePropertySourceEnvironmentPostProcessStrategy.java [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+
+        LinkedList<String> signingRegions = getSigningRegions(environment);
+        ListIterator<String> regionsIterator = signingRegions.listIterator(signingRegions.size());
+
+        // To keep the order of precedence, we iterate from the last one to the first one because of the method 'addFirst'.
+        // If we want the first region specified to be the first property source, we have to add it last.
+        while (regionsIterator.hasPrevious()) {
+            // We only want to halt boot (if true) for the last property source so the first one added.
+            boolean shouldHaltBoot = haltBoot && !regionsIterator.hasNext();
+            String region = regionsIterator.previous();
```
[the remark this answers is not in the document]
Comment (mcnoreau12) [reply src/main/java/com/coveo/configuration/parameterstore/strategy/MultiRegionParameterStorePropertySourceEnvironmentPostProcessStrategy.java]: It does not give me the result I want this way 😕 I want to know if the current region is the last one in the list. I'm iterating from the last one to the first one so when I check if there's a previous, all have a previous except for the first one but that't not the one I want...

File: src/main/java/com/coveo/configuration/parameterstore/strategy/MultiRegionParameterStorePropertySourceEnvironmentPostProcessStrategy.java [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+
+            environment.getPropertySources()
+                       .addFirst(buildParameterStorePropertySource(PARAMETER_STORE_PROPERTY_SOURCE_NAME + region,
```
[the remark this answers is not in the document]
Comment (mcnoreau12) [reply src/main/java/com/coveo/configuration/parameterstore/strategy/MultiRegionParameterStorePropertySourceEnvironmentPostProcessStrategy.java]: true. Done

File: src/main/java/com/coveo/configuration/parameterstore/strategy/ParameterStorePropertySourceEnvironmentPostProcessStrategyFactoryImpl.java [hunk 9]
```diff
@@ -0,0 +1,28 @@
...
+
+import java.util.HashMap;
+import java.util.Map;
+
+import com.amazonaws.regions.DefaultAwsRegionProviderChain;
+
+public class ParameterStorePropertySourceEnvironmentPostProcessStrategyFactoryImpl
+        implements ParameterStorePropertySourceEnvironmentPostProcessStrategyFactory
+{
+    public static final String DEFAULT_STRATEGY = "Default";
```
[the remark this answers is not in the document]
Comment (mcnoreau12) [reply src/main/java/com/coveo/configuration/parameterstore/strategy/ParameterStorePropertySourceEnvironmentPostProcessStrategyFactoryImpl.java]: done.

File: src/main/java/com/coveo/configuration/parameterstore/ParameterStorePropertySourceEnvironmentPostProcessor.java [hunk 10]
```diff
@@ -1,67 +1,61 @@
...
-    static final String PARAMETER_STORE_ENABLED_CONFIGURATION_PROPERTY = "awsParameterStorePropertySource.enabled";
-    static final String PARAMETER_STORE_HALT_BOOT_CONFIGURATION_PROPERTY = "awsParameterStorePropertySource.haltBoot";
-    static final String PARAMETER_STORE_CLIENT_ENDPOINT_CONFIGURATION_PROPERTY = "awsParameterStoreSource.ssmClient.endpointConfiguration.endpoint";
-    static final String PARAMETER_STORE_CLIENT_ENDPOINT_SIGNING_REGION_CONFIGURATION_PROPERTY = "awsParameterStoreSource.ssmClient.endpointConfiguration.signingRegion";
-    static final String PARAMETER_STORE_SUPPORT_MULTIPLE_APPLICATION_CONTEXTS_CONFIGURATION_PROPERTY = "awsParameterStorePropertySource.supportMultipleApplicationContexts";
-
-    private static final String PARAMETER_STORE_PROPERTY_SOURCE_NAME = "AWSParameterStorePropertySource";
-
     static boolean initialized;
+    static ParameterStorePropertySourceEnvironmentPostProcessStrategyFactory postProcessStrategyFactory = new ParameterStorePropertySourceEnvironmentPostProcessStrategyFactoryImpl();
```
[the remark this answers is not in the document]
Comment (mcnoreau12) [reply src/main/java/com/coveo/configuration/parameterstore/ParameterStorePropertySourceEnvironmentPostProcessor.java]: It's needed for tests like `initialized`.

File: src/main/java/com/coveo/configuration/parameterstore/ParameterStorePropertySourceConfigurationProperty.java [hunk 11]
```diff
@@ -0,0 +1,19 @@
...
+
+public final class ParameterStorePropertySourceConfigurationProperty
+{
+    private static final String PREFIX = "awsParameterStorePropertySource.";
+    private static final String SSM_CLIENT_PREFIX = "awsParameterStoreSource.ssmClient.";
+
+    public static final String ACCEPTED_PROFILE = "awsParameterStorePropertySourceEnabled";
+
+    public static final String ENABLED = PREFIX + "enabled";
+    public static final String ACCEPTED_PROFILES = PREFIX + "enabledProfiles";
```
Comment (fredboutin) [inline src/main/java/com/coveo/configuration/parameterstore/ParameterStorePropertySourceConfigurationProperty.java]: I would use String.join("." everywhere, instead of having the "." in the constants.

File: src/main/java/com/coveo/configuration/parameterstore/ParameterStorePropertySourceConfigurationProperty.java [hunk 12]
```diff
@@ -0,0 +1,19 @@
+package com.coveo.configuration.parameterstore;
+
+public final class ParameterStorePropertySourceConfigurationProperty
```
Comment (fredboutin) [inline src/main/java/com/coveo/configuration/parameterstore/ParameterStorePropertySourceConfigurationProperty.java]: ` ParameterStorePropertySourceConfigurationProperties` I feel like this class should have methods that return the values and even hold some logic to know if the lib is enabled for example. Doing this would simplify the post processors class and encapsulate the whole logic for configuration properties in here.

File: readme.md [hunk 13]
```diff
@@ -81,6 +81,12 @@ Spring Cloud has a second application context named bootstrap that gets initiali
 
 If you still want the post processor to run twice or if you are using [spring-boot-devtools](https://docs.spring.io/spring-boot/docs/current/reference/html/using-spring-boot.html#using-boot-devtools-restart), you can set the optional property `awsParameterStorePropertySource.supportMultipleApplicationContexts` to `true`. The default property value is `false`to prevent multiple initializations. If you are also using Spring Cloud, this property will only work if set in the bootstrap properties.
 
+## Multi-region
```
[the remark this answers is not in the document]
Comment (mcnoreau12) [reply readme.md]: done

File: readme.md (hunk 5 above)
Comment (mcnoreau12) [reply readme.md]: done

Comment (fredboutin) [inline readme.md]: How does this play with the already existing signingRegion parameter? is there some restrictions like not having both set at the same time? Even more important: are those enforced in the code?

File: src/main/java/com/coveo/configuration/parameterstore/strategy/MultiRegionParameterStorePropertySourceEnvironmentPostProcessStrategy.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (fredboutin) [reply src/main/java/com/coveo/configuration/parameterstore/strategy/MultiRegionParameterStorePropertySourceEnvironmentPostProcessStrategy.java]: the last one is the first you add, why not simply change the haltBoot value to false at the end of the while so it is true (or not depending on initial state) only on the first iteration?

File: src/main/java/com/coveo/configuration/parameterstore/strategy/DefaultParameterStorePropertySourceEnvironmentPostProcessStrategy.java [hunk 14]
```diff
@@ -0,0 +1,71 @@
...
+import com.amazonaws.client.builder.AwsClientBuilder.EndpointConfiguration;
+import com.amazonaws.regions.AwsRegionProviderChain;
+import com.amazonaws.services.simplesystemsmanagement.AWSSimpleSystemsManagement;
+import com.amazonaws.services.simplesystemsmanagement.AWSSimpleSystemsManagementClientBuilder;
+
+import com.coveo.configuration.parameterstore.ParameterStorePropertySource;
+import com.coveo.configuration.parameterstore.ParameterStorePropertySourceConfigurationProperty;
+import com.coveo.configuration.parameterstore.ParameterStoreSource;
+
+public class DefaultParameterStorePropertySourceEnvironmentPostProcessStrategy
```
Comment (fredboutin) [inline src/main/java/com/coveo/configuration/parameterstore/strategy/DefaultParameterStorePropertySourceEnvironmentPostProcessStrategy.java]: I would rename all the classes in this package since they are called in a post processor but it does not really matter here. What they really do is configure the property source. I'd replace the words `EnvironmentPostProcess` for `Configuration`.

Comment (fredboutin) [reply src/main/java/com/coveo/configuration/parameterstore/strategy/DefaultParameterStorePropertySourceEnvironmentPostProcessStrategy.java]: For example here it would be `DefaultParameterStorePropertySourceConfigurationStrategy `

File: src/main/java/com/coveo/configuration/parameterstore/ParameterStorePropertySourceConfigurationProperty.java [hunk 15]
```diff
@@ -0,0 +1,19 @@
...
+
+    public static final String ACCEPTED_PROFILE = "awsParameterStorePropertySourceEnabled";
+
+    public static final String ENABLED = PREFIX + "enabled";
+    public static final String ACCEPTED_PROFILES = PREFIX + "enabledProfiles";
+    public static final String HALT_BOOT = PREFIX + "haltBoot";
+    public static final String SUPPORT_MULTIPLE_APPLICATION_CONTEXTS = PREFIX + "supportMultipleApplicationContexts";
+    public static final String SSM_CLIENT_ENDPOINT_CONFIG_ENDPOINT = SSM_CLIENT_PREFIX
+            + "endpointConfiguration.endpoint";
+    public static final String SSM_CLIENT_ENDPOINT_CONFIG_SIGNING_REGION = SSM_CLIENT_PREFIX
```
[the remark this answers is not in the document]
Comment (fredboutin) [reply src/main/java/com/coveo/configuration/parameterstore/ParameterStorePropertySourceConfigurationProperty.java]: it's under endpointConfiguration (see the actual value), I prefer it this way to make a clear distinction between the two features

File: src/main/java/com/coveo/configuration/parameterstore/strategy/MultiRegionParameterStorePropertySourceEnvironmentPostProcessStrategy.java [hunk 16]
```diff
@@ -0,0 +1,62 @@
...
+
+    @Override
+    public void postProcess(ConfigurableEnvironment environment)
+    {
+        boolean haltBoot = environment.getProperty(HALT_BOOT, Boolean.class, Boolean.FALSE);
+
+        LinkedList<String> signingRegions = getSigningRegions(environment);
+        ListIterator<String> regionsIterator = signingRegions.listIterator(signingRegions.size());
+
+        // To keep the order of precedence, we iterate from the last one to the first one because of the method 'addFirst'.
```
Comment (fredboutin) [inline src/main/java/com/coveo/configuration/parameterstore/strategy/MultiRegionParameterStorePropertySourceEnvironmentPostProcessStrategy.java]: While you're at it, explain why you can't use the addLast method, it's clear for me, but my not be for newcomers.

File: src/main/java/com/coveo/configuration/parameterstore/strategy/ParameterStorePropertySourceEnvironmentPostProcessStrategy.java [hunk 17]
```diff
@@ -0,0 +1,8 @@
+package com.coveo.configuration.parameterstore.strategy;
+
+import org.springframework.core.env.ConfigurableEnvironment;
+
+public interface ParameterStorePropertySourceEnvironmentPostProcessStrategy
+{
+    void postProcess(ConfigurableEnvironment environment);
```
Comment (fredboutin) [inline src/main/java/com/coveo/configuration/parameterstore/strategy/ParameterStorePropertySourceEnvironmentPostProcessStrategy.java]: Following my rename comment  I'd rename this method to configureParameterStorePropertySources(...)

File: src/main/java/com/coveo/configuration/parameterstore/strategy/ParameterStorePropertySourceEnvironmentPostProcessStrategyFactory.java [hunk 18]
```diff
@@ -0,0 +1,6 @@
+package com.coveo.configuration.parameterstore.strategy;
+
+public interface ParameterStorePropertySourceEnvironmentPostProcessStrategyFactory
```
[the remark this answers is not in the document]
Comment (fredboutin) [reply src/main/java/com/coveo/configuration/parameterstore/strategy/ParameterStorePropertySourceEnvironmentPostProcessStrategyFactory.java]: I had the same comment, I don't really like the Impl suffix either.

[the remark this answers is not in the document]
Comment (mcnoreau12) [reply src/main/java/com/coveo/configuration/parameterstore/strategy/ParameterStorePropertySourceEnvironmentPostProcessStrategyFactory.java]: ditched.

File: src/main/java/com/coveo/configuration/parameterstore/strategy/MultiRegionParameterStorePropertySourceEnvironmentPostProcessStrategy.java [hunk 19]
```diff
@@ -0,0 +1,62 @@
...
+    }
+
+    private AWSSimpleSystemsManagement buildSSMClient(String region)
+    {
+        return AWSSimpleSystemsManagementClientBuilder.standard().withRegion(region).build();
+    }
+
+    private LinkedList<String> getSigningRegions(ConfigurableEnvironment environment)
+    {
+        String[] signingRegions = environment.getProperty(SSM_CLIENT_SIGNING_REGIONS, String[].class);
```
[the remark this answers is not in the document]
Comment (fredboutin) [reply src/main/java/com/coveo/configuration/parameterstore/strategy/MultiRegionParameterStorePropertySourceEnvironmentPostProcessStrategy.java]: I was going to say exactly this, also there is a test missing for an empty list (assert it throws after changing the behaviour)

File: readme.md (hunk 5 above)
Comment (fredboutin) [reply readme.md]: After reading the code I found out, but it should be written here that they are mutually exclusive

File: src/main/java/com/coveo/configuration/parameterstore/strategy/MultiRegionParameterStorePropertySourceEnvironmentPostProcessStrategy.java (hunk 19 above)
[the remark this answers is not in the document]
Comment (mcnoreau12) [reply src/main/java/com/coveo/configuration/parameterstore/strategy/MultiRegionParameterStorePropertySourceEnvironmentPostProcessStrategy.java]: done

File: src/main/java/com/coveo/configuration/parameterstore/strategy/DefaultParameterStorePropertySourceEnvironmentPostProcessStrategy.java [hunk 20]
```diff
@@ -0,0 +1,69 @@
...
+    private ParameterStorePropertySource buildParameterStorePropertySource(AWSSimpleSystemsManagement ssmClient,
+                                                                           boolean haltBoot)
+    {
+        return new ParameterStorePropertySource(PARAMETER_STORE_PROPERTY_SOURCE_NAME,
+                                                new ParameterStoreSource(ssmClient, haltBoot));
+    }
+
+    private AWSSimpleSystemsManagement buildSSMClient(ConfigurableEnvironment environment)
+    {
+        if (hasCustomEndpoint(environment)) {
```
Comment (mcnoreau12) [inline src/main/java/com/coveo/configuration/parameterstore/strategy/DefaultParameterStorePropertySourceEnvironmentPostProcessStrategy.java]: @fredboutin I was thinking about encapsulating this in another strategy. What do you think?

**Comment by mcnoreau12** [comment]
> Did you manually test this manually with aws? For both the multi-region and the default? Specifically the logic to fallback on multiregion.

Yes I did test it locally with aws on 3 differents regions for the default and mutli-region. I'm not sure what you mean by fallback on multiregion. It does not fallback on multi region but on default when `awsParameterStoreSource.multiRegion.ssmClient.regions` is not provided.

File: src/main/java/com/coveo/configuration/parameterstore/strategy/MultiRegionParameterStorePropertySourceEnvironmentPostProcessStrategy.java [hunk 21]
```diff
@@ -0,0 +1,63 @@
...
+                                                   Boolean.FALSE);
+
+        List<String> regions = getRegions(environment);
+        Collections.reverse(regions);
+
+        String lastRegion = regions.get(0);
+        environment.getPropertySources().addFirst(buildParameterStorePropertySource(lastRegion, haltBoot));
+
+        regions.stream()
+               .skip(1)
```
Comment (jebeaudet) [inline src/main/java/com/coveo/configuration/parameterstore/strategy/MultiRegionParameterStorePropertySourceEnvironmentPostProcessStrategy.java]: TIL about skip, nice

File: src/main/java/com/coveo/configuration/parameterstore/strategy/DefaultParameterStorePropertySourceEnvironmentPostProcessStrategy.java (hunk 20 above)
Comment (jebeaudet) [reply src/main/java/com/coveo/configuration/parameterstore/strategy/DefaultParameterStorePropertySourceEnvironmentPostProcessStrategy.java]: Not sure it's worth it IMO, at least for now

File: src/main/java/com/coveo/configuration/parameterstore/ParameterStorePropertySourceEnvironmentPostProcessor.java [hunk 22]
```diff
@@ -17,8 +21,10 @@
     static final String PARAMETER_STORE_ACCEPTED_PROFILES_CONFIGURATION_PROPERTY = "awsParameterStorePropertySource.enabledProfiles";
```
[the remark this answers is not in the document]
Comment (mcnoreau12) [reply src/main/java/com/coveo/configuration/parameterstore/ParameterStorePropertySourceEnvironmentPostProcessor.java]: done

File: readme.md (hunk 5 above)
Comment (mcnoreau12) [reply readme.md]: done

File: src/main/java/com/coveo/configuration/parameterstore/strategy/MultiRegionParameterStorePropertySourceConfigurationStrategy.java [hunk 23]
```diff
@@ -0,0 +1,69 @@
...
+    {
+        return AWSSimpleSystemsManagementClientBuilder.standard().withRegion(region).build();
+    }
+
+    private List<String> getRegions(ConfigurableEnvironment environment)
+    {
+        List<String> regions = CollectionUtils.arrayToList(environment.getProperty(ParameterStorePropertySourceConfigurationProperties.MULTI_REGION_SSM_CLIENT_REGIONS,
+                                                                                   String[].class));
+
+        if (CollectionUtils.isEmpty(regions)) {
```
Comment (fredboutin) [inline src/main/java/com/coveo/configuration/parameterstore/strategy/MultiRegionParameterStorePropertySourceConfigurationStrategy.java]: this if is not tested

File: src/main/java/com/coveo/configuration/parameterstore/ParameterStorePropertySourceEnvironmentPostProcessor.java [hunk 24]
```diff
@@ -5,63 +5,54 @@
...
-    static final String PARAMETER_STORE_ENABLED_CONFIGURATION_PROPERTY = "awsParameterStorePropertySource.enabled";
-    static final String PARAMETER_STORE_HALT_BOOT_CONFIGURATION_PROPERTY = "awsParameterStorePropertySource.haltBoot";
-    static final String PARAMETER_STORE_CLIENT_ENDPOINT_CONFIGURATION_PROPERTY = "awsParameterStoreSource.ssmClient.endpointConfiguration.endpoint";
-    static final String PARAMETER_STORE_CLIENT_ENDPOINT_SIGNING_REGION_CONFIGURATION_PROPERTY = "awsParameterStoreSource.ssmClient.endpointConfiguration.signingRegion";
-    static final String PARAMETER_STORE_SUPPORT_MULTIPLE_APPLICATION_CONTEXTS_CONFIGURATION_PROPERTY = "awsParameterStorePropertySource.supportMultipleApplicationContexts";
-
-    private static final String PARAMETER_STORE_PROPERTY_SOURCE_NAME = "AWSParameterStorePropertySource";
-
     static boolean initialized;
+    protected static ParameterStorePropertySourceConfigurationStrategyFactory postProcessStrategyFactory = new ParameterStorePropertySourceConfigurationStrategyFactory();
```
[the remark this answers is not in the document]
Comment (mcnoreau12) [reply src/main/java/com/coveo/configuration/parameterstore/ParameterStorePropertySourceEnvironmentPostProcessor.java]: done

File: src/main/java/com/coveo/configuration/parameterstore/ParameterStorePropertySourceEnvironmentPostProcessor.java [hunk 25]
```diff
@@ -5,63 +5,54 @@
...
-                                         Boolean.FALSE)) {
+            getPostProcessStrategy(environment).configureParameterStorePropertySources(environment);
+
+            if (doesNotSupportMultipleApplicationContexts(environment)) {
                 initialized = true;
             }
         }
     }
 
+    private ParameterStorePropertySourceConfigurationStrategy getPostProcessStrategy(ConfigurableEnvironment environment)
```
[the remark this answers is not in the document]
Comment (mcnoreau12) [reply src/main/java/com/coveo/configuration/parameterstore/ParameterStorePropertySourceEnvironmentPostProcessor.java]: done

File: src/main/java/com/coveo/configuration/parameterstore/strategy/MultiRegionParameterStorePropertySourceConfigurationStrategy.java (hunk 23 above)
Comment (mcnoreau12) [reply src/main/java/com/coveo/configuration/parameterstore/strategy/MultiRegionParameterStorePropertySourceConfigurationStrategy.java]: done

File: src/main/java/com/coveo/configuration/parameterstore/ParameterStorePropertySourceEnvironmentPostProcessor.java (hunk 25 above)
[the remark this answers is not in the document]
Comment (fredboutin) [reply src/main/java/com/coveo/configuration/parameterstore/ParameterStorePropertySourceEnvironmentPostProcessor.java]: hum the name is the same? or you havent push yet?
